build_tools/azure/install.sh

```diff
@@ -59,84 +48,31 @@ pre_python_environment_install() {
     fi
 }
 
-python_environment_install() {
-    if [[ "$DISTRIB" == "conda" || "$DISTRIB" == *"mamba"* ]]; then
-
-        if [[ "$CONDA_CHANNEL" != "" ]]; then
-            TO_INSTALL="--override-channels -c $CONDA_CHANNEL"
-        else
-            TO_INSTALL=""
-        fi
-
-        if [[ "$DISTRIB" == *"pypy"* ]]; then
-            TO_INSTALL="$TO_INSTALL pypy"
-        else
-            TO_INSTALL="$TO_INSTALL python=$PYTHON_VERSION"
-        fi
-
-        TO_INSTALL="$TO_INSTALL ccache pip blas[build=$BLAS]"
-
-        TO_INSTALL="$TO_INSTALL $(get_dep numpy $NUMPY_VERSION)"
-        TO_INSTALL="$TO_INSTALL $(get_dep scipy $SCIPY_VERSION)"
-        TO_INSTALL="$TO_INSTALL $(get_dep cython $CYTHON_VERSION)"
-        TO_INSTALL="$TO_INSTALL $(get_dep joblib $JOBLIB_VERSION)"
-        TO_INSTALL="$TO_INSTALL $(get_dep pandas $PANDAS_VERSION)"
-        TO_INSTALL="$TO_INSTALL $(get_dep pyamg $PYAMG_VERSION)"
-        TO_INSTALL="$TO_INSTALL $(get_dep Pillow $PILLOW_VERSION)"
-        TO_INSTALL="$TO_INSTALL $(get_dep matplotlib $MATPLOTLIB_VERSION)"
-
-        if [[ "$UNAMESTR" == "Darwin" ]] && [[ "$SKLEARN_TEST_NO_OPENMP" != "true" ]]; then
-                TO_INSTALL="$TO_INSTALL compilers llvm-openmp"
-        fi
-
-        make_conda $TO_INSTALL
+python_environment_install_and_activate() {
+    if [[ "$DISTRIB" == "conda"* ]]; then
+        conda update -n base conda -y
+        # pin conda-lock to latest released version (needs manual update from time to time)
+        conda install -c conda-forge conda-lock==1.0.5 -y
+        conda-lock install --name $VIRTUALENV $LOCK_FILE
+        source activate $VIRTUALENV
 
-    elif [[ "$DISTRIB" == "ubuntu" ]] || [[ "$DISTRIB" == "debian-32" ]]; then
+    elif [[ "$DISTRIB" == "ubuntu" || "$DISTRIB" == "debian-32" ]]; then
         python3 -m virtualenv --system-site-packages --python=python3 $VIRTUALENV
         source $VIRTUALENV/bin/activate
+        pip install -r "${LOCK_FILE}"
+    fi
 
-        python -m pip install $(get_dep cython $CYTHON_VERSION) \
-                $(get_dep joblib $JOBLIB_VERSION)
-
-    elif [[ "$DISTRIB" == "conda-pip-latest" ]]; then
-        # Since conda main channel usually lacks behind on the latest releases,
-        # we use pypi to test against the latest releases of the dependencies.
-        # conda is still used as a convenient way to install Python and pip.
-        make_conda "ccache python=$PYTHON_VERSION"
-        python -m pip install -U pip
-
-        python -m pip install pandas matplotlib scikit-image pyamg
-        # do not install dependencies for lightgbm since it requires scikit-learn.
-        python -m pip install "lightgbm>=3.0.0" --no-deps
```

Comment:
I think we do `--no-deps` here to not install scikit-learn. Currently the lock file includes the latest release version of scikit-learn. Do you see an issue with this?

Reply:
Indeed I saw this. scikit-learn is pinned in the lock files (since this is a dependency of of lightgbm) but we install the scikit-learn development version on top of it. It is slightly weird but I don't think this is an issue and this has the merit of simplicity.

I think at one point I looked at whether we can used `--no-deps` in the environment.yml file. If I remember correctly this may be possible with `conda` (https://github.com/conda/conda/pull/3969) but for `conda-lock`, `poetry` is used for the locking and having the equivalent functionality in `poetry` was not straightforward.